### PR TITLE
Fix path issue and capture problem determination data for script execution

### PR DESCRIPTION
**Pull request by ptitzler**
### What changes were proposed in this pull request?
Based on research performed for #2073

For local pipeline execution:
- Capture invocation information to aid with troubleshooting if generic components fail to execute
- Fix path handling for Windows platform
- Remove no-op parameter from Python script execution invocation

Example output (user runs pipeline that processes an r script but r is not installed):

```
[E 2021-08-30 08:58:13.875 ServerApp] Error executing rrrrrr.r: [Errno 2] No such file or directory: 'Rscript': 'Rscript'
[I 2021-08-30 08:58:13.876 ServerApp] Data capture for previous error:
     command: Rscript /Users/patti/elyra_dev/i2073/workspace/rrrrrr.r
     working directory: /Users/patti/elyra_dev/i2073/workspace
     environment variables: {'TMPDIR': '/var/folders/h8/l3qh8v192rv_vvggc6wyphzc0000gn/T/', '__CFBundleIdentifier': 'com.apple.Terminal', 'XPC_FLAGS': '0x0', 'TERM': 'xterm-256color', 'SSH_AUTH_SOCK': '/private/tmp/com.apple.launchd.UPRFGPMBxC/Listeners', 'XPC_SERVICE_NAME': '0', 'TERM_PROGRAM': 'Apple_Terminal', 'TERM_PROGRAM_VERSION': '440', 'TERM_SESSION_ID': 'D7E10F30-16B5-4397-A607-1E9F8A19A84F', 'SHELL': '/bin/zsh', 'HOME': '/Users/patti', ...}
```

Example output (user runs pipeline that executes a Python script but the `python3` executable in the PATH -- simulated by changing the command to `python4` from `python3`):

```
[E 2021-09-02 07:48:35.677 ServerApp] Error executing a-script.py: [Errno 2] No such file or directory: 'python4': 'python4'
[I 2021-09-02 07:48:35.679 ServerApp] Data capture for previous error:
     command: python4 /Users/patti/elyra_dev/i2073/workspace/a-script.py
     working directory: /Users/patti/elyra_dev/i2073/workspace
     environment variables: {'TMPDIR': 
```

### How was this pull request tested?

(MacOS only - no access to Windows) Confirmed that pd information is logged in the following scenarios:
- Run R script in local environment where R is not installed
- Run Python script in local environment where `python3` is not in the PATH

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: elyra/pipeline/processor_local.py [hunk 1]
```diff
@@ -269,6 +269,8 @@ def process(self, operation: GenericOperation, elyra_run_name: str):
         except CalledProcessError as cpe:
             error_msg = str(cpe.stderr.decode())
             self.log.error(f'Error executing {file_name}: {error_msg}')
+            # Log process information to aid with troubleshooting
+            self.log.info(f'FFDC: \n argv: {argv}\n cwd: {file_dir}\n env:{envs}')
```
[the remark this answers is not in the document]
Comment (ptitzler) [reply elyra/pipeline/processor_local.py]: See updated message text. I also added code that properly logs this information on other platforms, where the error is not raised via `CalledProcessError`.

**Review by kevin-bates — APPROVED** [review]
LGTM. Thanks Patrick.

**Comment by kevin-bates** [comment]
> Run Python script in local environment where python3 is not in the PATH

I don't see any accommodations for a missing `python3` executable.

**Comment by ptitzler** [comment]
> > Run Python script in local environment where python3 is not in the PATH
> 
> I don't see any accommodations for a missing `python3` executable.

I updated the description to properly reflect what has been tested. The PR does not attempt to resolve the problem scenario that `python3` is not accessible.

**Comment by kevin-bates** [comment]
Thanks for the update. I think we can still run into an issue on Windows here. If we want to defer that portion of the issue (which _was_ the issue) I'm okay so long as another issue is created. 
I'm also okay updating this issue and can probably look into a fix later today if that's warranted.

**Comment by ptitzler** [comment]
> If we want to defer that portion of the issue (which was the issue) I'm okay so long as another issue is created.

Even though I didn't state that explicitly, that's the reason why this PR does not close the issue using one of the magic keyword(s).

I didn't attempt to fix that specific problem because
- A basic platform dependent solution (use `python` on win, `python3` elsewhere)  would likely not be sufficient because (1) there are Python installations on windows that install `python3` and (2) `python` is not guaranteed to be at version 3, which is why we changed the executable name in the first place.
- I don't have an environment to verify that a potential generic proposal (try `python3` followed by `python`) would work as expected.

I also looked into updating the documentation to state what the prerequisites are but there's currently no topic that would lend itself.

**Comment by kevin-bates** [comment]
I'm in the process of provisioning a Windows server to take a closer look at the python vs. python3 issue.

**Comment by kevin-bates** [comment]
This is interesting.  Although python3.exe can be found using where, and that directory is in the PATH, it is not found from the active conda env, and the active conda env doesn't include a python3.exe:
```
(base) C:\Users\Administrator>echo %PATH%
C:\Users\Administrator\anaconda3;C:\Users\Administrator\anaconda3\Library\mingw-w64\bin;C:\Users\Administrator\anaconda3\Library\usr\bin;C:\Users\Administrator\anaconda3\Library\bin;C:\Users\Administrator\anaconda3\Scripts;C:\Users\Administrator\anaconda3\bin;C:\Users\Administrator\anaconda3\condabin;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0;C:\WINDOWS\System32\OpenSSH;C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps;.

(base) C:\Users\Administrator>where python
C:\Users\Administrator\anaconda3\python.exe
C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps\python.exe

(base) C:\Users\Administrator>where python3
C:\Users\Administrator\AppData\Local\Microsoft\WindowsApps\python3.exe

(base) C:\Users\Administrator>python --version
Python 3.8.8

(base) C:\Users\Administrator>python3 --version
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.
```

I'd like to see how [`shutil.which`](https://docs.python.org/3.3/library/shutil.html#shutil.which) behaves here and that might be a simple solution.  Can't get to this today though..

**Comment by kevin-bates** [comment]
In looking into `shutil.which`, I think the solution, for all platforms, is to simply use the python executable that is currently running the jupyter server: `sys.executable`.  We know that _that_ python is an appropriate version and available.

This may need to be revisited if we ever wanted to introduce _local runtime configurations_ (which I think we may want to look into someday), but, for local processing right now its arguably the most correct thing we can do.

File: elyra/pipeline/processor_local.py [hunk 2]
```diff
@@ -288,7 +300,7 @@ class PythonScriptOperationProcessor(ScriptOperationProcessor):
     _script_type = 'Python'
 
     def get_argv(self, file_path) -> List[str]:
-        return ['python3', file_path, '--PYTHONHOME', os.path.dirname(file_path)]
+        return ['python3', file_path]
```
Comment (kevin-bates) [inline elyra/pipeline/processor_local.py:303]: Per this comment, I think we should go with `sys.executable` here.
```suggestion
        return [f"{sys.executable}", file_path]
```

File: elyra/pipeline/processor_local.py [hunk 3]
```diff
@@ -23,6 +23,7 @@
 import time
 from typing import Dict
 from typing import List
+from typing import Optional
```
Comment (kevin-bates) [inline elyra/pipeline/processor_local.py:26]: `sys` is needed for the `sys.executable` reference below.
```suggestion
from typing import Optional
import sys
```

**Comment by kevin-bates** [comment]
I've successfully tested the use of `sys.executable` on Windows (having removed python3.exe).  The local unit test will exercise for Linux, so I think we're good to go.

**Pull request merged**